tendermint/src/proposal.rs

```diff
@@ -0,0 +1,267 @@
+//! Proposals from validators
+
+mod canonical_proposal;
+mod msg_type;
+mod sign_proposal;
+
+pub use self::canonical_proposal::CanonicalProposal;
+pub use msg_type::Type;
+pub use sign_proposal::{SignProposalRequest, SignedProposalResponse};
+
+use crate::block::{Height, Id as BlockId, Round};
+use crate::chain::Id as ChainId;
+use crate::consensus::State;
+use crate::Signature;
+use crate::Time;
+use crate::{Error, Kind};
+use bytes::BufMut;
+use std::convert::{TryFrom, TryInto};
+use tendermint_proto::types::Proposal as RawProposal;
+use tendermint_proto::{DomainType, Error as DomainTypeError};
+
+/// Proposal
+#[derive(Clone, PartialEq, Debug)]
+pub struct Proposal {
+    /// Proposal message type
+    pub msg_type: Type,
+    /// Height
+    pub height: Height,
+    /// Round
+    pub round: Round,
+    /// POL Round
+    pub pol_round: Option<Round>,
+    /// Block ID
+    pub block_id: Option<BlockId>,
+    /// Timestamp
+    pub timestamp: Option<Time>,
+    /// Signature
+    pub signature: Signature,
+}
+
+impl DomainType<RawProposal> for Proposal {}
+
+impl TryFrom<RawProposal> for Proposal {
+    type Error = Error;
+
+    fn try_from(value: RawProposal) -> Result<Self, Self::Error> {
+        if value.pol_round < -1 {
+            return Err(Kind::NegativePOLRound.into());
+        }
+        let pol_round = match value.pol_round {
+            -1 => None,
+            n => Some(Round::try_from(n)?),
+        };
+        Ok(Proposal {
+            msg_type: value.r#type.try_into()?,
+            height: value.height.try_into()?,
+            round: value.round.try_into()?,
+            pol_round,
+            block_id: match value.block_id {
+                None => None,
+                Some(raw_block_id) => Some(BlockId::try_from(raw_block_id).unwrap()),
+            },
+            timestamp: match value.timestamp {
+                None => None,
+                Some(t) => Some(t.try_into()?),
+            },
```

Comment:
This can probably be rewritten as follows (though that might obscure the meaning a bit for people not familiar with [`Option::transpose`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose) so feel free to disregard):

```suggestion
            timestamp: value.timestamp.map(|t| t.try_into()).transpose()?,
```

Reply:
Covered in fixes #3 commit.